### ISSUE TITLE
Implement permittedUnfreePackages

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#55674]
## Issue description

**allowUnfreePredicate** is too complicated for the most common usecase: allow individual unfree packages.

```
  nixpkgs.config = {
    allowUnfreePredicate = (pkg: builtins.elem (builtins.parseDrvName pkg.name).name [ "flashplayer" "vscode" ]);
    firefox = {
      # constantly broken https://github.com/NixOS/nixpkgs/issues/55657
      #enableAdobeFlash = true;
    };
  };
```

I suggest to add **permittedUnfreePackages** (name similar to permittedInsecurePackages) as discussed in https://github.com/NixOS/nixpkgs/issues/49774:

```
  nixpkgs.config = {
    permittedUnfreePackages = [ "flashplayer" "vscode" ];
  };
```

Document it as the suggested way for this usecase and add a proper example for allowUnfreePredicate where it's needed.

### Steps to reproduce

See `configuration.nix` as a former Ansible user and think it's way too complicated.

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.98, NixOS, 18.09.2137.168cbb39691 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(davidak): `""`
 - channels(root): `"nixos-18.09.2137.168cbb39691, nixos-hardware, nixos-unstable-19.03pre169108.36f31600749"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by 7c6f434c** [comment]
Might be a good idea. If you get around to making a PR, feel free to mention me for review and (after some time for other comments) merge.

**Comment by FRidh** [comment]
What if both `allowUnfreePredicate` and `permittedUnfreePackages` are specified? What should happen?

**Comment by 7c6f434c** [comment]
The insecure case sets a precedent that predicate overrides package list.

We might consider saying that the predicate and the package list are joined via «or» (I hope we make the same decision everywhere)

**Comment by oxij** [comment]
#44518 :)

However, in my local branch I now just set the default predicate to compare via a regexp, because comparing by name is not enough sometimes, e.g. drivers/firmwares usually list both version and kernel version in the name and sometimes you want to whitelist only a specific firmware version for all kernel versions.

**Comment by 7c6f434c** [comment]
Yes, this is also why the example for the predicate uses `hasPrefix`.

@matthewbauer maybe #44518 would indeed simplify things for newcomers.

**Comment by ShamrockLee** [comment]
This function is still important to me!

**Comment by ShamrockLee** [comment]
> meanwhile, this works
> 
> ```nix
> {
> # allow ALL unfree packages
> #nixpkgs.config.allowUnfree = true;
> 
> # allow some unfree packages
> nixpkgs.config.allowUnfreePredicate = pkg: builtins.elem (lib.getName pkg) [
>   "osu-lazer"
>   "flashplayer"
>   "vscode"
>   "cnijfilter2" # canon printer
>   "font-bh-lucidatypewriter-75dpi" # https://github.com/NixOS/nixpkgs/issues/99014
>   "corefonts"
> ];
> }
> ```
> 
> via https://discourse.nixos.org/t/cant-make-allowunfreepredicate-in-configuration-nix/10882/3

There's an issue with this approach:
Ifthe internal package splitting / wrapping of an unfree package (e.g. steam) changes, there would be another unfree package with a different name that doesn't got covered by the list.

Of course we can use something like `builtins.match` (which is what I use in my configuration), but that's ambiguous and stringy.

**Comment by ptman** [comment]
@ShamrockLee Is it really necessary to block on fixing hypothetical future problems? If package names change, then configuration has to change to match.

**Comment by zseri** [comment]
also, this usually won't be a silent failure either, because if the affected packages are pulled in, the evaluation would fail because they weren't marked as permitted yet.

**Comment by Majiir** [comment]
A use case for a list-based option is declaring unfree packages in multiple modules. Setting `allowUnfreePredicate` multiple times in different modules does not compose. In my testing, one of the predicates "wins" and the others are (silently!) ignored. A list would allow definitions from multiple modules to merge and form a single predicate.

> We might consider saying that the predicate and the package list are joined via «or» (I hope we make the same decision everywhere)

+1

**Comment by bb2020** [comment]
> A use case for a list-based option is declaring unfree packages in multiple modules. Setting `allowUnfreePredicate` multiple times in different modules does not compose. In my testing, one of the predicates "wins" and the others are (silently!) ignored. A list would allow definitions from multiple modules to merge and form a single predicate.

+1

**Comment by landure** [comment]
I post to unstaled this ticket. 

In [Installing insecure packages @ Nixpkgs documentation](https://nixos.org/manual/nixpkgs/stable/#sec-allow-insecure),
it's possible to allow an insecure package with:

```nix
{ permittedInsecurePackages = [ "hello-1.2.3" ]; }
```

This allows to write something like:

```nix
{ permittedInsecurePackages = [ (builtins.toString pkgs.hello) ]; }
```

According to [Installing unfree packages @ Nixpkgs documentation](https://nixos.org/manual/nixpkgs/stable/#sec-allow-unfree), there's no equivalent option for unfree package.

A possible solution to this ticket is adding the `permittedUnfreePackages` option:

```nix
{ permittedUnfreePackages = [ "hello-1.2.3" ]; }
```

This allows to write something like:

```nix
{ permittedUnfreePackages = [ (builtins.toString pkgs.hello) ]; }
```

This is a more readable alternative to `allowUnfreePredicate`, and it has the benefit that it works for permitted unfree packages defined in separates files.

A solution to the `allowUnfreePredicate` problem is given on the linked discourse. See the code below.
(source: <https://codeberg.org/AndrewKvalheim/configuration/src/commit/11794e595144500a6c2be706e42ed698b1788bb8/packages/nixpkgs-issue-55674.nix>)

```nix
{ config, lib, ... }:
let
  inherit (builtins) elem;
  inherit (lib) getName mkOption;
  inherit (lib.types) listOf str;
in
{
  # Pending https://github.com/NixOS/nixpkgs/issues/55674
  options.allowedUnfree = mkOption { type = listOf str; default = [ ]; };
  config.nixpkgs.config.allowUnfreePredicate = p: elem (getName p) config.allowedUnfree;
}
```
